lua/weapons/grav_blaster.lua

```diff
@@ -0,0 +1,232 @@
+	fx:SetOrigin(burst.pos)
+
+	util.Effect("Explosion", fx)
+
+	-- Toss all props inwards and players away in the explosion radius
+	local entities = ents.FindInSphere(burst.pos, BURST_EXPLOSION_RADIUS_UNITS)
+
+	for _, entity in ipairs(entities) do
+		applyGravityToEntity(burst, entity, BURST_TOSS_IMPULSE_MAGNITUDE)
+	end
+end
+
+local function updateBurst(burst)
+	local timeElapsed = CurTime() - burst.timeCreated
+	if timeElapsed > BURST_LIFETIME_SECS then
+		return true
+	end
+
+	-- check if the burst has collided
+	local endPos = burst.pos + burst.dir * BURST_SPEED_UNITS * timeElapsed
+	local trace = util.TraceLine({
+		start = burst.pos,
+		endpos = endPos,
+		filter = {
+			burst.owner,
+		},
+	})
+
+	if trace and trace.Hit then
+		return true
+	end
+
+	local entities = ents.FindInSphere(endPos, BURST_PASSIVE_RADIUS_UNITS)
+	for _, entity in ipairs(entities) do
+		if entity == burst.owner then
+			continue
+		end
+
+		applyGravityToEntity(burst, entity, BURST_PASSIVE_IMPULSE_MAGNITUDE)
+	end
+
+	burst.pos = endPos
+
+	return false
+end
+
+function SWEP:FireBurst(position, direction)
+	if CLIENT then
+		return
+	end
+
+	table.insert(globalBursts, makeBurst(position, direction, self:GetOwner(), self))
+end
+
+function SWEP:Initialize()
+	self:SetHoldType("physgun")
+end
+
+function SWEP:Reload()
+	if self:GetOwner():GetAmmoCount(self.Primary.Ammo) == 0 then
+		return
+	end
+
+	if self.Reloading then
+		return
+	end
+
+	self:EmitSound(self.LoadSound)
+	self.Reloading = true
+
+	-- Wait until the sound has finished playing
+	timer.Simple(self.LoadSoundDurationSecs, function()
+		self.Reloading = false
+		if not IsValid(self) or not IsValid(self:GetOwner()) then
+			return
+		end
+
+		local ammo = self:GetOwner():GetAmmoCount(self.Primary.Ammo)
+
+		self:SetClip1(math.min(ammo, self.Primary.ClipSize))
+		self:GetOwner():RemoveAmmo(self.Primary.ClipSize, self.Primary.Ammo)
+	end)
+
+	-- 0.1 is arbitrary because apparently SetClip1 doesn't update in the same tick,
+	-- so the user could reload in the same tick that the reload finished
+	self:SetNextPrimaryFire(CurTime() + self.LoadSoundDurationSecs + 0.1)
+end
+
+function SWEP:CalcViewModelView(_, _, _, newPos, newAng)
+	-- Gangsta wielding
+	return newPos, Angle(0, 0, -15) + newAng
+end
+
+function SWEP:PrimaryAttack()
+	if not self:CanPrimaryAttack() then
+		return
+	end
+
+	self:SetNextPrimaryFire(CurTime() + 1 / self.PrimaryRateOfFire)
+	self:FireBurst(self:GetOwner():GetShootPos(), self:GetOwner():GetAimVector())
+	self:EmitSound(self.ShootSound)
+	self:TakePrimaryAmmo(1)
+	self:GetOwner():ViewPunch(Angle(-3, -0.6, -0.3))
+	self:SendWeaponAnim(ACT_VM_PRIMARYATTACK)
+end
+
+function SWEP:SecondaryAttack() end
+
+if SERVER then
+	hook.Add("Think", "tas.grav_blaster_update", function()
+		for index, burst in ipairs(globalBursts) do
+			renderBurst(burst)
+			local burstDied = updateBurst(burst)
+
+			if burstDied then
+				explodeBurst(burst)
+				table.remove(globalBursts, index)
```

Comment:
This is going to break the loop.

Think storing globalBursts as a lookup table and using `pairs` instead would work.

Reply:
What would be the key? can tables be serialized reliably

Reply:
Just store as a hashset with the burst as the key, and yes